File: reference/js/plan.js
"use strict";
// Staleness planner: turns a fresh manifest plus the previously written index
// into the two work sets the build runs.
//
//   planFetch    what to pull from SQL Server, by modify_date
//   planDescribe what to send to the model, by fingerprint
//
// Both are pure. The fetch stage runs between them, and the split is what makes
// a rebuild after a release cost a few dozen describe calls instead of 1237:
// the release moves modify_date on everything it touched, and the fingerprint
// then discards the ones whose content is unchanged.
//
// An object missing from `state` is new and always both fetched and described.
// An object in `state` but absent from the manifest was dropped from the
// database; planFetch reports it so the writer can remove its index row.

const { fingerprint, sampleDepth } = require("./fingerprint.js");

function keyOf(object) {
  return `${object.schema}.${object.name}`;
}

// Reasons are recorded per object so a build log can say why it did the work.
//
// `sample` applies to tables only. A table is described partly from rows, so
// growth can make the description stale on its own; a procedure is described
// from its body, which modify_date already covers. Scoping the rule to tables
// says that, where letting it run on every kind only worked by accident —
// procedures report zero rows, so it never fired.
const FETCH_REASONS = [
  { reason: "new", kinds: null, when: (object, prior) => !prior },
  { reason: "modified", kinds: null, when: (object, prior) => prior.modified !== object.modified },
  {
    reason: "sample",
    kinds: ["table"],
    when: (object, prior) => sampleDepth(prior.rows) !== sampleDepth(object.rows),
  },
];

function applies(rule, object) {
  return rule.kinds === null || rule.kinds.includes(object.kind);
}

// state: Map<key, { modified, fingerprint, rows?, description? }>
// Returns { fetch, reuse, dropped, reasons }.
function planFetch(objects, state = new Map()) {
  const fetch = [];
  const reuse = [];
  const reasons = {};
  const seen = new Set();

  for (const object of objects) {
    const key = keyOf(object);
    seen.add(key);
    const prior = state.get(key);
    const hit = FETCH_REASONS.find((r) => applies(r, object) && r.when(object, prior));
    if (hit) {
      reasons[key] = hit.reason;
      fetch.push(object);
    } else {
      reuse.push(object);
    }
  }

  const dropped = [...state.keys()].filter((key) => !seen.has(key));
  return { fetch, reuse, dropped, reasons };
}

// `fetched` is [{ object, payload, ... }] straight from the fetch stage:
// payload is whatever fingerprint() needs for that kind. Every other field on
// an entry is carried through untouched, because the describe stage needs the
// structure and key that came with it. Objects whose content hash matches the
// index keep their existing description and cost nothing.
function planDescribe(fetched, state = new Map()) {
  const describe = [];
  const unchanged = [];

  for (const item of fetched) {
    const current = fingerprint(item.object.kind, item.payload);
    const prior = state.get(keyOf(item.object));
    const entry = { ...item, fingerprint: current };
    if (prior && prior.fingerprint === current && prior.description) unchanged.push(entry);
    else describe.push(entry);
  }

  return { describe, unchanged };
}

// Only objects scope marks describable reach the model; a synonym is indexed
// for its target, not for a generated sentence about it.
function describable(entries) {
  return entries.filter((e) => e.object.describe !== false);
}

function summarize({ fetch, reuse, dropped }, describePlan) {
  return {
    fetch: fetch.length,
    reuse: reuse.length,
    dropped: dropped.length,
    describe: describePlan ? describePlan.describe.length : undefined,
    unchanged: describePlan ? describePlan.unchanged.length : undefined,
  };
}

module.exports = { keyOf, planFetch, planDescribe, describable, summarize, FETCH_REASONS };

File: internal/plan/plan.go
// Package plan turns a fresh manifest plus the previously written index into
// the two work sets a build runs: Fetch, decided by modify signal, and
// Describe, decided by content fingerprint. Both are pure; the fetch stage runs
// between them. The split is what makes a rebuild after a release cost a few
// dozen describe calls rather than one per object — the release moves the
// modify signal on everything it touched, and the fingerprint then discards the
// ones whose content is unchanged.
//
// An object missing from the prior state is new and is always both fetched and
// described. An object in the state but absent from the manifest was dropped
// from the database; Fetch reports it so the writer can remove its index row.
package plan

import (
	"sort"

	"github.com/branow/dbmap/internal/catalog"
	"github.com/branow/dbmap/internal/fingerprint"
)

// Reason records why an object is being refetched, so a build log can say what
// it is doing and why.
type Reason string

const (
	// New is an object the previous build never saw.
	New Reason = "new"
	// Modified is an object whose engine modify signal moved, or that carries
	// no modify signal at all and therefore can never be proven untouched.
	Modified Reason = "modified"
	// Sample is a table whose row count crossed the depth the describer sees.
	Sample Reason = "sample"
)

// Rule is one row of the fetch table.
type Rule struct {
	Reason Reason
	// Kinds scopes a rule to particular kinds; nil means every kind.
	Kinds []catalog.Kind
	// When is asked only after every earlier rule declined, so every rule but
	// the first may assume prior is non-nil.
	When func(object catalog.Object, prior *catalog.State) bool
}

// FetchReasons is the fetch decision as data, in the order it is asked; the
// first rule that applies and fires names the reason. The sample rule is scoped
// to tables because only a table is described partly from its rows — on other
// kinds it would be a no-op that merely looks intentional.
var FetchReasons = []Rule{
	{
		Reason: New,
		When:   func(_ catalog.Object, prior *catalog.State) bool { return prior == nil },
	},
	{
		Reason: Modified,
		When: func(object catalog.Object, prior *catalog.State) bool {
			return !object.Modified.Same(prior.Modified)
		},
	},
	{
		Reason: Sample,
		Kinds:  []catalog.Kind{catalog.Table},
		When: func(object catalog.Object, prior *catalog.State) bool {
			return fingerprint.SampleDepth(prior.Rows) != fingerprint.SampleDepth(object.Rows)
		},
	},
}

func (r Rule) applies(object catalog.Object) bool {
	if r.Kinds == nil {
		return true
	}
	for _, kind := range r.Kinds {
		if kind == object.Kind {
			return true
		}
	}
	return false
}

// FetchPlan is what the fetch stage is asked to do, and what it may skip.
type FetchPlan struct {
	Fetch []catalog.Object
	Reuse []catalog.Object
	// Dropped are keys the previous build indexed that the database no longer
	// has, sorted, so the writer can remove their rows.
	Dropped []string
	// Reasons maps an object key to why it is being fetched.
	Reasons map[string]Reason
}

// Fetch decides what to pull from the database, given the state the previous
// build left in the index it wrote.
func Fetch(objects []catalog.Object, state map[string]catalog.State) FetchPlan {
	plan := FetchPlan{Reasons: make(map[string]Reason)}
	seen := make(map[string]struct{}, len(objects))

	for _, object := range objects {
		key := object.Key()
		seen[key] = struct{}{}
		var prior *catalog.State
		if recorded, ok := state[key]; ok {
			prior = &recorded
		}
		if reason, hit := reasonFor(object, prior); hit {
			plan.Reasons[key] = reason
			plan.Fetch = append(plan.Fetch, object)
			continue
		}
		plan.Reuse = append(plan.Reuse, object)
	}

	for key := range state {
		if _, ok := seen[key]; !ok {
			plan.Dropped = append(plan.Dropped, key)
		}
	}
	sort.Strings(plan.Dropped)
	return plan
}

func reasonFor(object catalog.Object, prior *catalog.State) (Reason, bool) {
	for _, rule := range FetchReasons {
		if rule.applies(object) && rule.When(object, prior) {
			return rule.Reason, true
		}
	}
	return "", false
}

// DescribePlan is what the model is asked to look at, and what it is not.
type DescribePlan struct {
	Describe []catalog.Entry
	// Unchanged are entries whose content hash matches the index. They keep the
	// description already written for them and cost nothing.
	Unchanged []catalog.Entry
}

// Describe splits freshly fetched entries by content fingerprint, each returned
// carrying the hash just computed so the writer never recomputes one. An entry
// is redescribed when its hash moved, or when the hash matches but the prior
// state holds no description to reuse.
func Describe(fetched []catalog.Entry, state map[string]catalog.State) (DescribePlan, error) {
	var plan DescribePlan

	for _, entry := range fetched {
		current, err := fingerprint.Of(entry.Object, entry.Structure)
		if err != nil {
			return DescribePlan{}, err
		}
		entry.Fingerprint = current

		prior, ok := state[entry.Key()]
		if ok && prior.Fingerprint == current && prior.Description != "" {
			entry.Description = prior.Description
			plan.Unchanged = append(plan.Unchanged, entry)
			continue
		}
		plan.Describe = append(plan.Describe, entry)
	}

	return plan, nil
}

// Describable keeps only the entries whose kind earns a generated sentence.
func Describable(entries []catalog.Entry) []catalog.Entry {
	var kept []catalog.Entry
	for _, entry := range entries {
		if entry.Object.Describable() {
			kept = append(kept, entry)
		}
	}
	return kept
}

// Summary is the one-line shape of a plan, for a build's progress output.
type Summary struct {
	Fetch     int
	Reuse     int
	Dropped   int
	Describe  int
	Unchanged int
}

// Summarize counts both plans together.
func Summarize(fetch FetchPlan, describe DescribePlan) Summary {
	return Summary{
		Fetch:     len(fetch.Fetch),
		Reuse:     len(fetch.Reuse),
		Dropped:   len(fetch.Dropped),
		Describe:  len(describe.Describe),
		Unchanged: len(describe.Unchanged),
	}
}
